test/integration/infra.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/golang/glog"
+
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	proxyconfig "istio.io/api/proxy/v1/config"
+	"istio.io/manager/cmd"
+	"istio.io/manager/platform/kube/inject"
+	"istio.io/manager/test/util"
+)
+
+type infra struct {
+	// docker tags
+	Hub, Tag   string
+	MixerImage string
+	CaImage    string
+
+	Namespace string
+	Verbosity int
+
+	// map from app to pods
+	apps map[string][]string
+
+	Auth proxyconfig.ProxyMeshConfig_AuthPolicy
+
+	// switches for infrastructure components
+	Mixer   bool
+	Ingress bool
+	Egress  bool
+
+	namespaceCreated bool
+}
+
+func (infra *infra) setup() error {
+	if infra.Namespace == "" {
+		var err error
+		if infra.Namespace, err = util.CreateNamespace(client); err != nil {
+			return err
+		}
+		infra.namespaceCreated = true
+	} else {
+		if _, err := client.Core().Namespaces().Get(infra.Namespace, meta_v1.GetOptions{}); err != nil {
+			return err
+		}
+	}
+
+	deploy := func(name string) error {
+		if yaml, err := fill(name, infra); err != nil {
+			return err
+		} else if err = infra.kubeApply(yaml); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	if err := deploy("config.yaml.tmpl"); err != nil {
+		return err
+	}
+	if err := deploy("manager.yaml.tmpl"); err != nil {
+		return err
+	}
+	if err := deploy("mixer.yaml.tmpl"); err != nil {
+		return err
+	}
+	if infra.Auth != proxyconfig.ProxyMeshConfig_NONE {
+		if err := deploy("ca.yaml.tmpl"); err != nil {
+			return err
+		}
+	}
+	if infra.Ingress {
+		if err := deploy("ingress-proxy.yaml.tmpl"); err != nil {
+			return err
+		}
+	}
+	if infra.Egress {
+		if err := deploy("egress-proxy.yaml.tmpl"); err != nil {
+			return err
+		}
+	}
+
```

Comment:
I think you had suggested making this table based before. Might be a good time to do that now, e.g.

```golang
deployments := []struct {
        filename string
        deploy   bool
    }{
        {"config.yaml.tmpl", true},
        {"manager.yaml.tmpl", true},
        {"mixer.yaml.tmpl", true},
        {"ca.yaml.tmpl", infra.Auth != proxyconfig.ProxyMeshConfig_NONE},
        {"ingress-proxy.yaml.tmpl", infra.Ingress},
        {"egress-proxy.yaml.tmpl", infra.Egress},
    }
    for _, d := range deployments {
        if d.deploy {
            if err := d.deploy(d.filename); err != nil {
                return err
            }
        }
    }
```